sdk/eventhub/event-hubs/src/eventDataBatch.ts

```diff
@@ -251,8 +271,27 @@ export class EventDataBatchImpl implements EventDataBatch {
    * Generates an AMQP message that contains the provided encoded events and annotations.
    * @param encodedEvents - The already encoded events to include in the AMQP batch.
    * @param annotations - The message annotations to set on the batch.
+   * @param publishingProps - Idempotent publishing properties used to decorate the events in the batch while sending.
    */
-  private _generateBatch(encodedEvents: Buffer[], annotations?: MessageAnnotations): Buffer {
+  private _generateBatch(
+    encodedEvents: Buffer[],
+    annotations: MessageAnnotations | undefined,
+    publishingProps?: PartitionPublishingProperties
+  ): Buffer {
+    if (this._isIdempotent && publishingProps) {
+      // We need to decode the encoded events, add the idempotent annotations, and re-encode them.
+      // We can't lazily encode the events because we rely on `message.encode` to capture the
+      // byte length of anything not in `event.body`.
+      // Events can't be decorated ahead of time because the publishing properties aren't known
+      // until the events are being sent to the service.
+      const decodedEvents = encodedEvents.map(message.decode) as unknown as RheaMessage[];
+      const decoratedEvents = this._decorateRheaMessagesWithPublishingProps(
```

Comment:
why is it called decorated?

Reply:
I would have to ask Chris :) I guess it means adding something on top of rhea message (decorator pattern?)